### PR TITLE
fltk GUI toolkit

### PR DESCRIPTION
**Pull request by alen12345**
Needed by gmsh especially

### CODE REVIEW

**Comment by alen12345** [comment]
It is not actually passing tests. I will write a final post when I am done.

**Comment by alen12345** [comment]
All tests are passing really now!

**Comment by alen12345** [comment]
No, tests on Windows are not...!

**Comment by alen12345** [comment]
@CJ-Wright why aren't test builds on Windows in AppVeyor triggered?

**Comment by alen12345** [comment]
@CJ-Wright I have fixed all tests; can you please have a last look?

File: recipes/fltk/build.sh [hunk 1]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+mkdir build && cd build
+cmake -DCMAKE_INSTALL_PREFIX=$PREFIX -DCMAKE_BUILD_TYPE=Release ..
+make
```
Comment (isuruf) [inline recipes/fltk/build.sh]: Add `-j${CPU_COUNT}`

File: recipes/fltk/build.sh [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+mkdir build && cd build
+cmake -DCMAKE_INSTALL_PREFIX=$PREFIX -DCMAKE_BUILD_TYPE=Release ..
```
Comment (isuruf) [inline recipes/fltk/build.sh]: How about building shared libraries? `-DBUILD_SHARED_LIBS=yes`

File: recipes/fltk/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  number: 0
+
+requirements:
+  build:
+    - toolchain
+    - cmake
+
+test:
+  commands:
+    - echo "Tests have already been executed after compilation."
```
Comment (isuruf) [inline recipes/fltk/meta.yaml]: Can you add,
```
- conda inspect linkages -p $PREFIX fltk  # [not win]
- conda inspect objects -p $PREFIX fltk  # [osx]
```

**Comment by alen12345** [comment]
@isuruf thanks a lot for the suggestions! I have included all your suggestions!

File: recipes/fltk/build.sh [hunk 4]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+mkdir build && cd build
+cmake -DCMAKE_INSTALL_PREFIX=$PREFIX -DCMAKE_BUILD_TYPE=Release .. -DBUILD_SHARED_LIBS=yes
```
Comment (isuruf) [inline recipes/fltk/build.sh]: This should have been `-DOPTION_BUILD_SHARED_LIBS=yes`. Sorry about the mistake

File: recipes/fltk/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  url: http://fltk.org/pub/{{ name }}/{{ version }}/{{ name }}-{{ longversion }}-source.tar.bz2
+  md5: {{ md5 }}
+
+build:
+  number: 0
+
+requirements:
+  build:
+    - toolchain
+    - cmake
```
Comment (isuruf) [inline recipes/fltk/meta.yaml]: Can you add `zlib, jpeg, png` to build and run requirements, so that it doesn't use bundled dependencies?

**Comment by isuruf** [comment]
Can you add `-DCMAKE_INSTALL_LIBDIR=lib` so that the libraries are installed in `${PREFIX}/lib`?

Also could you add this patch, so that the library name is sane? (i.e. `libfltk.so` instead of `libfltk_SHARED.so`)
```
--- CMake/macros.cmake.old      2018-02-26 00:54:06.391674683 -0600
+++ CMake/macros.cmake  2018-02-26 00:54:19.327396916 -0600
@@ -38,7 +38,7 @@
 
     set_target_properties(${LIBRARY_NAME}
         PROPERTIES
-        OUTPUT_NAME ${LIBRARY_NAME}
+        OUTPUT_NAME ${LIBNAME}
         DEBUG_OUTPUT_NAME ${LIBRARY_NAME_DEBUG}
         CLEAN_DIRECT_OUTPUT TRUE
         COMPILE_DEFINITIONS "FL_LIBRARY"
```

File: recipes/fltk/meta.yaml (hunk 5 above)
Comment (isuruf) [reply recipes/fltk/meta.yaml]: You'll have to pin those libraries as well.

**Comment by alen12345** [comment]
@isuruf please guide me here because I am not so experienced in these build commands.

File: recipes/fltk/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    - zlib
+    - jpeg
+    - libpng
```
Comment (isuruf) [inline recipes/fltk/meta.yaml]: Change these to `libpng >=1.6.32,<1.6.35`, `zlib 1.2.11`, `jpeg 9*` and the same in run?

File: recipes/fltk/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  run:
+    - zlib
+    - jpeg
+    - libpng
+
+test:
+  commands:
+    - echo "Tests have already been executed after compilation."
```
Comment (isuruf) [inline recipes/fltk/meta.yaml]: Can you add tests like, https://github.com/conda-forge/bzip2-feedstock/blob/master/recipe/meta.yaml#L47-L55 ?

Comment (alen12345) [reply recipes/fltk/meta.yaml]: Yes sure next commit :)

**Comment by isuruf** [comment]
Can you also add lines like https://github.com/conda-forge/jpeg-feedstock/blob/master/recipe/meta.yaml#L14-L28 so that the library is built for different versions?

**Comment by alen12345** [comment]
@isuruf still issue on windows; please tell me what to change for that too, so that I can also add the other suggestions in the next commit.

**Comment by isuruf** [comment]
Doing this https://github.com/conda-forge/staged-recipes/pull/5192#issuecomment-368409794 should fix your windows issues

**Comment by alen12345** [comment]
@isuruf still failing after excluding [win and py36]; am I missing something from your suggestions? Tests will be implemented later on.

**Pull request closed without merging**